Feature/packaging

mitzimorris opened the pull request:
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

- Replaced CmdStan install script `make_cmdstan.sh` with Python script `bin/install_cmdstah.py`- (issue https://github.com/stan-dev/cmdstanpy/issues/41)

- Updated `setup.py` to make `install_cmdstan` available as command line script with package install

- Tested package upload and download via TestPyPI - works on my machine (issue https://github.com/stan-dev/cmdstanpy/issues/17)

- Updated docs to use ScikitLearn templates and styling.

- Edited docs and ipython notebook to reflect current API


#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):  Columbia University


By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+   -v, --version : version, defaults to latest
+   -d, --dir : install directory, defaults to '~/.cmdstanpy
+"""
+import argparse
```
argparse works. Click makes it a bit easier to create "manual"

mitzimorris replied:
went with argparse because it's part of the standard library this script has a simple enough set of command line options.

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+import argparse
+import contextlib
+import os
+import os.path
```
This is not needed, `os` imports `os.path` by default

mitzimorris replied:
thanks - will clean up!

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+import subprocess
+import sys
+import tarfile
+import urllib.request
```
Is there a difference with `request`? Is urllib the default? (vs urllib2, urllib3)

mitzimorris replied:
for me (Mac OS, Python 3.7.0) there is a difference w/ request - import of `urllib` gives the following error:

> AttributeError: module 'urllib' has no attribute 'request'

ahartikainen replied:
Yes, the import format is correct

I was thinking https://2.python-requests.org/en/master/

But I think urllib is the correct choice here.

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+        cmd = ['make', 'build']
+        proc = subprocess.Popen(
+            cmd, cwd=None, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
```
Would `subprocess.call` work better?

mitzimorris replied:
need to capture stderr if make fails.

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+        sys.exit(3)
+    print('Download successful, file: {}'.format(file_tmp))
+    try:
+        tar = tarfile.open(file_tmp)
```
Does `with tarfile.open ... as tar` work? Or do this need `finally` to close the file?

mitzimorris replied:
using `finally` to close.

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+    print('Download successful, file: {}'.format(file_tmp))
+    try:
+        tar = tarfile.open(file_tmp)
+        tar.extractall(os.getcwd())
```
Is there some progressbar option?

mitzimorris replied:
progressbar option as part of `extractall` function?  no.

agreed that we need progress bars for download, untar, and compile - will file an issue.

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+        if proc.returncode:
+            print('Command "make build" failed')
+            if stderr:
+                print(stderr.decode('ascii').strip())
```
`utf-8`?

mitzimorris replied:
good catch!  changed to utf-8 everywhere.

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
```
Could this file have `.py` extension? It still can be called without `.py` on linux

mitzimorris replied:
doesn't seem to work on Mac OS.

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+
+
+def main():
+    parser = argparse.ArgumentParser()
```
Should this output help/manual if no arguments are given?

mitzimorris replied:
if no arguments are given the latest version is installed in `~/.cmdstanpy`.
it only prints the usage if there's an unrecognized argument.

it's in the documentation, FWIW.

ahartikainen commented on bin/install_cmdstan:
```diff
@@ -0,0 +1,164 @@
+                print(stderr.decode('ascii').strip())
+            sys.exit(3)
+        print('Test model compilation')
+        cmd = ['make', os.path.join('examples', 'bernoulli', 'bernoulli')]
```
On some systems make is not called make. Maybe it could try to use environmental var with default `make`

mitzimorris replied:
this will be a problem for the `compile_model` function as well.
needs more investigation - will file another issue.

mitzimorris commented:
@ahartikainen - thanks for your careful review!

if you think I've addressed all issues, will merge.
I've tested the dist on TestPyPI - do you think this is ready for upload to PyPI?

ahartikainen commented:
LGTM
